### PR TITLE
add OnClickRow to also supoort onSelect

### LINKED ISSUE #5451: DataTable — cannot do `onSelect` and `onClickRow` together
### Expected Behavior

The clickable section of "onClickRow" should not include the selection checkbox so that the user could click on row content and also successfully check/uncheck the box.

### Actual Behavior

The clickable section of the row includes the checkbox, so a user cannot select a checkbox without the `onClickRow` function from being executed.

### URL, screen shot, or Codepen exhibiting the issue

https://codesandbox.io/s/datatable-onselectonclickrow-7384x?file=/index.js

### Steps to Reproduce

1. Click on Checkbox in a body row
2. The `onClickRow` function is triggered, so user can't make a selection

### Your Environment

- Grommet version: 2.17.3
- Browser Name and version: MacOS Chrome
- Operating System and version (desktop or mobile): MacOS Mojave 10.14.6

### PR DESCRIPTION
**Pull request by britt6612**
#### What does this PR do?
This PR adds a option for the user to have the whole row clickable when using `onSelect` 
#### Where should the reviewer start?
DataTable/body.js
#### What testing has been done on this PR?
storybook
#### How should this be manually tested?
storybook 
Added `onClickRow="onSelect"` to the story
#### Do Jest tests follow these best practices?

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `userEvent` is used in place of `fireEvent`.
- [ ] `asFragment()` is used for snapshot testing.

#### Any background context you want to provide?
It was brought up that users were not able to do onSelect and onClickRow together 
we decided this would not be something we recommend. If using onSelect then your onClickRow should be set to `onSelect` if user is wanting the whole row to be clickable otherwise `onClickRow` should not be used.
#### What are the relevant issues?
Closes #5451 

#### Do the grommet docs need to be updated?
yes
#### Should this PR be mentioned in the release notes?
yes
#### Is this change backwards compatible or is it a breaking change?
changed the keyboard behavior

### CODE REVIEW

File: src/js/components/DataTable/Body.js [hunk 1]
```diff
@@ -45,34 +45,30 @@ const Row = memo(
...
+          if (onClickRow && !isDisabled) {
+            if (typeof onClickRow === 'function') {
+              // extract from React's synthetic event pool
+              event.persist();
+              const adjustedEvent = event;
+              adjustedEvent.datum = datum;
+              adjustedEvent.index = index;
+              onClickRow(adjustedEvent);
+            }
+            if (onClickRow === 'onSelect') {
```
Comment (taysea) [inline src/js/components/DataTable/Body.js:58]:
```suggestion
            else if (onClickRow === 'onSelect') {
```
liking need to fix formatting locally w linter.

File: src/js/components/DataTable/Body.js [hunk 2]
```diff
@@ -45,34 +45,30 @@ const Row = memo(
...
-                setActive(index);
-              }
-            : undefined
+          onClickRow && !isDisabled ? () => setActive(index) : undefined
         }
         onMouseLeave={onClickRow ? () => setActive(undefined) : undefined}
       >
         {(selected || onSelect) && (
           <Cell
-            background={
```
Comment (taysea) [inline src/js/components/DataTable/Body.js:71]: Why was this removed?

File: src/js/components/DataTable/Body.js [hunk 3]
```diff
@@ -181,20 +178,25 @@ const Body = forwardRef(
...
-                event.persist();
-                const adjustedEvent = event;
-                adjustedEvent.datum = data[active];
-                onClickRow(adjustedEvent);
-              }
-            : undefined
-        }
+        onEnter={(event) => {
+          if (
+            (onClickRow && active >= 0 && !disabled) ||
```
Comment (taysea) [inline src/js/components/DataTable/Body.js:183]: Formatting seems to have changed here. It should be:

```
(onClickRow && active >= 0 && (!disabled || !disabled.includes(datumValue(data[active], primaryProperty))

File: src/js/components/DataTable/Body.js [hunk 4]
```diff
@@ -181,20 +178,25 @@ const Body = forwardRef(
...
+            !disabled.includes(datumValue(data[active], primaryProperty))
+          ) {
+            if (typeof onClickRow === 'function') {
+              event.persist();
+              const adjustedEvent = event;
+              adjustedEvent.datum = data[active];
+              onClickRow(adjustedEvent);
+            }
+            if (onClickRow === 'onSelect') {
```
Comment (taysea) [inline src/js/components/DataTable/Body.js:192]:
```suggestion
            else if (onClickRow === 'onSelect') {
```
Fix formatting locally w linter

File: src/js/components/DataTable/stories/typescript/OnSelect.tsx [hunk 5]
```diff
@@ -25,16 +25,9 @@ const columns: ColumnConfig<RowType>[] = [
...
-          pad={{ horizontal: 'small', vertical: 'xsmall' }}
-          onClick={() => {}}
-          hoverIndicator
-          fill
-        >
-          <Text truncate weight="bold">
-            {datum.location}
-          </Text>
-        </Box>
+        <Text truncate weight="bold">
```
Comment (taysea) [inline src/js/components/DataTable/stories/typescript/OnSelect.tsx:28]: Any reason we changed this?

Comment (britt6612) [reply src/js/components/DataTable/stories/typescript/OnSelect.tsx:28]: took this out so we can test how it is without having a row clickable since we wouldn't recommend doing this I can put back in

File: src/js/components/DataTable/Body.js (hunk 4 above)
Comment (britt6612) [reply src/js/components/DataTable/Body.js]: done

File: src/js/components/DataTable/Body.js (hunk 3 above)
Comment (britt6612) [reply src/js/components/DataTable/Body.js]: fixed

File: src/js/components/DataTable/Body.js (hunk 2 above)
Comment (britt6612) [reply src/js/components/DataTable/Body.js:71]: put back in shouldn't have been removed

File: src/js/components/DataTable/Body.js (hunk 1 above)
Comment (britt6612) [reply src/js/components/DataTable/Body.js]: done

File: src/js/components/DataTable/Body.js [hunk 6]
```diff
@@ -181,20 +182,26 @@ const Body = forwardRef(
...
-            !disabled.includes(datumValue(data[active], primaryProperty)))
-            ? (event) => {
-                event.persist();
-                const adjustedEvent = event;
-                adjustedEvent.datum = data[active];
-                onClickRow(adjustedEvent);
-              }
-            : undefined
-        }
+        onEnter={(event) => {
```
Comment (taysea) [inline src/js/components/DataTable/Body.js:185]: The WCAG recommendation for checkboxes is to select them with "Space" (https://www.w3.org/TR/wai-aria-practices-1.1/examples/checkbox/checkbox-1/checkbox-1.html). That is how datatable keyboard selection currently works. So this would be a breaking change to only support "Enter" for selection.

Proposing that we support "Space" also for the selection when `onClickRow="onSelect"`.

Comment (britt6612) [reply src/js/components/DataTable/Body.js:185]: I'm glad you brought this up I was thinking the same especially since the normal behavior of `checkbox` is a "Space" bar I will add this with a comment

File: src/js/components/DataTable/Body.js [hunk 7]
```diff
@@ -181,20 +182,42 @@ const Body = forwardRef(
...
+              const primaryValue = data[active]?.[primaryProperty];
+              if (selected && selected.includes(primaryValue)) {
+                onSelect(selected.filter((s) => s !== primaryValue));
+              } else onSelect([...selected, primaryValue]);
+            }
+          }
+        }}
+        // The WCAG recommendation for checkboxes is to select them with "Space"
+        onSpace={() => {
+          if (
```
Comment (taysea) [inline src/js/components/DataTable/Body.js:207]: Instead of duplicating this code for `onEnter` and `onSpace`, is there any way to create reusable function that both use?

Comment (britt6612) [reply src/js/components/DataTable/Body.js:211]: yes I pulled this out but left the `If` statement for the first part

File: src/js/components/DataTable/Body.js [hunk 8]
```diff
@@ -179,22 +180,45 @@ const Body = forwardRef(
...
+            (!disabled ||
+              !disabled.includes(datumValue(data[active], primaryProperty)))
+          ) {
+            if (typeof onClickRow === 'function') {
+              event.persist();
+              const adjustedEvent = event;
+              adjustedEvent.datum = data[active];
+              onClickRow(adjustedEvent);
+            } else if (onClickRow === 'onSelect') {
+              selectCheckBox();
```
Comment (taysea) [inline src/js/components/DataTable/Body.js:205]: naming suggestion: `selectRow()`

File: src/js/components/DataTable/Body.js [hunk 9]
```diff
@@ -179,22 +180,45 @@ const Body = forwardRef(
...
-                event.persist();
-                const adjustedEvent = event;
-                adjustedEvent.datum = data[active];
-                onClickRow(adjustedEvent);
-              }
-            : undefined
-        }
+        onEnter={(event) => {
+          if (
+            onClickRow &&
```
Comment (taysea) [inline src/js/components/DataTable/Body.js:194]: Could we potentially pull this out into a variable and place before the return:

```
const clickedRow = onClickRow && active >= 0 &&
            (!disabled ||
              !disabled.includes(datumValue(data[active], primaryProperty)))
```

And then use that in the onEnter and onSpace as 

```
if (clickedRow) {
...
}

File: src/js/components/DataTable/Body.js (hunk 8 above)
Comment (britt6612) [reply src/js/components/DataTable/Body.js:205]: I like that better

**Review by taysea — APPROVED** [review]
LGTM!

File: src/js/components/DataTable/index.d.ts [hunk 10]
```diff
@@ -105,7 +105,9 @@ export interface DataTableProps<TRowType = any> {
   step?: number;
 
   // Events
-  onClickRow?: (event: MouseClick<TRowType> | KeyPress<TRowType>) => void;
+  onClickRow?:
+    | 'onSelect'
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/index.d.ts:109]: The term "onSelect" seems a little odd to me here. Like it will call any `onSelect` handler when the row is clicked. I'm wondering if we should use "select" here, indicating that if the user clicks the row, it will be treated as initiating a selection.

File: src/js/components/DataTable/Body.js [hunk 11]
```diff
@@ -179,22 +180,41 @@ const Body = forwardRef(
...
     const [lastActive, setLastActive] = React.useState();
 
+    const selectRow = () => {
+      const primaryValue = data[active]?.[primaryProperty];
+      if (selected && selected.includes(primaryValue)) {
+        onSelect(selected.filter((s) => s !== primaryValue));
+      } else onSelect([...selected, primaryValue]);
+    };
+
+    const clickedRow =
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/Body.js:190]: The term "clickedRow" seems to imply that the row was clicked. Perhaps "clickableRow" or "clickEnabled"?

File: src/js/components/DataTable/Body.js [hunk 12]
```diff
@@ -179,22 +180,41 @@ const Body = forwardRef(
...
-            !disabled.includes(datumValue(data[active], primaryProperty)))
-            ? (event) => {
-                event.persist();
-                const adjustedEvent = event;
-                adjustedEvent.datum = data[active];
-                onClickRow(adjustedEvent);
-              }
-            : undefined
-        }
+        onEnter={(event) => {
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/Body.js:198]: Can we only bind an event handler if the condition is true, rather than always binding a function and then testing whether it should do anything? For example: `onEnter={clickedRow ? (event) => { if ... } : undefined`

File: src/js/components/DataTable/Body.js [hunk 13]
```diff
@@ -45,24 +45,24 @@ const Row = memo(
...
-                // extract from React's synthetic event pool
-                event.persist();
-                const adjustedEvent = event;
-                adjustedEvent.datum = datum;
-                adjustedEvent.index = index;
-                onClickRow(adjustedEvent);
-              }
-            : undefined
-        }
+        onClick={(event) => {
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/Body.js:48]: Can we only bind a handler if needed? For example: `onClick={(onClickRow && !isDisabled) ? (event) => {} : undefined}`

File: src/js/components/DataTable/Body.js (hunk 11 above)
Comment (britt6612) [reply src/js/components/DataTable/Body.js:190]: I can see that I like `clickableRow`

File: src/js/components/DataTable/Body.js (hunk 13 above)
Comment (britt6612) [reply src/js/components/DataTable/Body.js]: done

File: src/js/components/DataTable/Body.js (hunk 12 above)
Comment (britt6612) [reply src/js/components/DataTable/Body.js]: done

File: src/js/components/DataTable/index.d.ts (hunk 10 above)
Comment (britt6612) [reply src/js/components/DataTable/index.d.ts]: that's fair and makes sense

**Review by ericsoderberghp — APPROVED** [review]
